web/packages/teleport/src/UnifiedResources/ResourceCard.tsx

```diff
@@ -54,60 +52,146 @@ const ResTypeIconBox = styled(Box)`
+  // of the component that contains all the labels. If this component is taller
+  // than the height of a single label row, we show a "+x more" button.
+  React.useEffect(() => {
+    if (!labelsInnerContainer.current) return;
+
+    const observer = new ResizeObserver(entries => {
+      const container = entries[0];
+
+      // We're taking labelRowHeight * 1.5 just in case some glitch adds or
+      // removes a pixel here and there.
+      const moreThanOneRow =
+        container.contentBoxSize[0].blockSize > labelRowHeight * 1.5;
+      setShowMoreLabelsButton(moreThanOneRow);
+
+      // Count number of labels in the first row. This will let us calculate and
+      // show the number of labels left out from the view.
+      const labelElements = [
+        ...entries[0].target.querySelectorAll('[data-is-label]'),
+      ];
+      const firstLabelPos = labelElements[0]?.getBoundingClientRect().top;
+      // Find the first one below.
+      const firstLabelInSecondRow = labelElements.findIndex(
+        e => e.getBoundingClientRect().top > firstLabelPos
+      );
+
+      setNumMoreLabels(
+        firstLabelInSecondRow > 0
+          ? labelElements.length - firstLabelInSecondRow
+          : 0
+      );
+    });
+
+    observer.observe(labelsInnerContainer.current);
+    return () => {
+      observer.disconnect();
+    };
+  });
+
+  const onMoreLabelsClick = () => {
+    setShowAllLabels(true);
+  };
+
   return (
-    <CardContainer p={3} alignItems="start">
-      <CheckboxInput type="checkbox" mx={0}></CheckboxInput>
-      <ResourceIcon
-        alignSelf="center"
-        name={resIcon}
-        width="45px"
-        height="45px"
-        ml={2}
-      />
-      {/* MinWidth is important to prevent descriptions from overflowing. */}
-      <Flex flexDirection="column" flex="1" minWidth="0" ml={3} gap={1}>
-        <Flex flexDirection="row" alignItems="start">
-          <SingleLineBox flex="1" title={name}>
-            <Text typography="h5">{name}</Text>
-          </SingleLineBox>
-          <ButtonBorder size="small">Connect</ButtonBorder>
-        </Flex>
-        <Flex flexDirection="row" alignItems="center">
-          <ResTypeIconBox>
-            <ResTypeIcon size={18} />
-          </ResTypeIconBox>
-          {description.primary && (
-            <SingleLineBox ml={1} title={description.primary}>
-              <Text typography="body2" color="text.slightlyMuted">
-                {description.primary}
-              </Text>
+    <CardContainer>
+      <CardInnerContainer
+        p={3}
+        alignItems="start"
+        showAllLabels={showAllLabels}
+        onMouseLeave={() => setShowAllLabels(false)}
+        // Class name needed to properly propagate hover state to child
+        // components. If we don't do it, there's no way to sync the background
+        // of the "more" button, since it's opaque and the background is
+        // animated.
+        className="grv-unified-resource-card"
```

Comment:
We should be able to get away with not using class names directly here, but it's fine for me if we leave it like this. In that case, maybe let's add a TODO comment to address this in the future.

https://stackoverflow.com/a/41010688/742872